### PR TITLE
Paper: alphagenome-pt: Flexible Training of AlphaGenome Models in PyTorch

### PR DESCRIPTION
**Pull request by RylieWeaver**
Pull request for SciPy 2026 proceedings submission.

--------------
If you are creating this PR in order to submit a draft of your paper, please name your PR with `Paper: <title>`. An editor will then add a `draft` label; this will trigger GitHub Actions to run automated checks on your paper and build a preview. You may then work to resolve failed checks and ensure the preview build looks correct. If you have any questions, please tag the proceedings team in a comment on your PR with `@scipy-conference/2026-proceedings`.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#preview-your-paper) for more information.

### CODE REVIEW

**Comment by KandiSanjana** [comment]
I will be a reviewer for this paper.

**Comment by RylieWeaver** [comment]
> I will be a reviewer for this paper.

Sounds good :)

I see that 3 checks are still failing because of missing DOIs. For these specific three, I did do a search and wasn't able to find the DOIs. Two are NeurIPS citations whose ArXiV preprints do have DOIs, in case it's better to have the ArXiV DOI than nothing? At the moment, I am leaving them as is, but please do let me know if I should adjust them somehow.

Thank you for taking the time to be a reviewer! @KandiSanjana

**Comment by anwarkhan-ai** [comment]
I have reviewed this paper as a 2nd reviewer @KandiSanjana, adding my comments here.

File: papers/rylie_weaver/main.md [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
...SpliceAI-Nature-2019; @Pangolin-PubMed-2022; @EVO2-Nature-2026; @EVO-nguyen-2024; @PlantCAD2-2025; @NucleotideTransformer-Nature-2025; @NucleotideTransformer-biorxiv-2023; @DNABERT-2-arxiv-2024; @DNABert-Bioinformatics-2021; @HyenaDNA-NeurIPS-2023]. Within this broader area, one important direction is to predict function directly from DNA sequence, which is a central challenge in genomics. AlphaGenome advances this direction by predicting a wide range of functional genomic outputs, including gene expression, chromatin accessibility, splicing, and chromatin contact maps, directly from DNA sequence for human and mouse genomes [@AlphaGenome-Nature-2026]. These predictions are valuable because they provide fast computational estimates for quantities that would otherwise require relatively expensive and time-consuming experiments. Moreover, DNA sequence inputs can be precisely specified and systematically modified in silico, enabling controlled perturbations that are difficult or impossible to perform experimentally. A sufficiently accurate and general model could accelerate progress in biology across major sectors such as healthcare, pharmaceuticals, agriculture, energy, biotechnology.
```
Comment (anwarkhan-ai) [inline papers/rylie_weaver/main.md]: I think there are 2 categories of modes, sequence to function (Enformer, Basenji...) and language models (Evo, HyenaDNA...) I will like them split. Minor but major sector impact is unverified.

File: papers/rylie_weaver/main.md [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+:::
+
+The model begins with an encoder that consists of seven convolutional downsampling blocks, which capture short-range dependencies and coarsen the sequence from 1bp to $2^7=128$bp resolution. The resulting 128bp embeddings are then input to transformer layers [@Transformers-NeurIPS-2017], which capture long-range dependencies over the full context at reduced sequence length. The transformer attention logits are adjusted by an attention bias computed from pairwise sequence representations at a coarser $(2048 \times 2048)$ bp resolution. These pairwise representations are updated with a pair-to-pair attention mechanism (called row attention in AlphaGenome) where each pair representation attends across its row, which improves the model's capability to capture higher-order relationships. Finally, AlphaGenome applies a decoder that consists of seven convolutional upsampling blocks, which capture short-range dependencies and restores the sequence 1 bp resolution. U-Net-style skip connections [@UNet-MICCAI-2015] with the fine-grained encoder representations are used to prevent information loss.
```
Comment (anwarkhan-ai) [inline papers/rylie_weaver/main.md]: Optional suggestion, "Coarsifies" in the figure should be coarsens

File: papers/rylie_weaver/main.md [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+
+The resulting 1bp, 128bp, and $(2048 \times 2048)$ bp representations are then passed to task-specific output heads. In total, the encoder-transformer-decoder structure allows AlphaGenome to retain fine-grained information while also incorporating long-range context.
+
+
+## Use Cases and Design Philosophy
```
Comment (anwarkhan-ai) [inline papers/rylie_weaver/main.md:64]: I think there should be a implementation section before package layout, Also a validation section that will help us better understand parameter count vs model, on a gpu along with throughput and memory footprint. Also if possible training curve.

File: papers/rylie_weaver/main.md [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+assert small_cfg.pair_channels / small_cfg.num_channels == official_cfg.pair_channels / official_cfg.num_channels == 1 / 6
+assert small_cfg.num_splice_sites / small_cfg.max_seq_len == official_cfg.num_splice_sites / official_cfg.max_seq_len == 1 / 2048
+assert small_cfg.splice_site_channels / small_cfg.num_channels == official_cfg.splice_site_channels / official_cfg.num_channels == 1
+```
+
+In addition to training from scratch, `alphagenome-pt` supports training from converted AlphaGenome checkpoints. The package provides utilities for instantiating a checkpoint-compatible public AlphaGenome configuration, converting the released JAX parameters to PyTorch, and loading those parameters into the PyTorch model. This is useful for users who want to finetune from released AlphaGenome weights rather than begin from random initialization. At the same time, checkpoint loading is still a current limitation of the package. At present, `alphagenome-pt` supports loading the all-folds model and folds 0, 1, 2, and 3, but we are still validating numerical equivalence with the original JAX implementation.
```
Comment (anwarkhan-ai) [inline papers/rylie_weaver/main.md]: I think its a very important question that validation is still in progress. test_mapping_state.py which compare the parameter mapping with JAX, Can you please share what is it showing ? all parameter mapped ? key per fold analysis ?

File: papers/rylie_weaver/main.md [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+    AlphaGenome,
+    load_alphagenome_checkpoint,
+    official_alphagenome_config,
+)
+
+# Instantiate full official model
+cfg = official_alphagenome_config()
+model = AlphaGenome(cfg)
+
+# Downloads from Hugging Face if the checkpoint is not already present
```
Comment (anwarkhan-ai) [inline papers/rylie_weaver/main.md]: which hugging face repo ? default package is RylieWeaver/alphagenome-pytorch ?

File: papers/rylie_weaver/myst.yml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    - deep learning
+  # Add the abbreviations that you use in your paper here
+  abbreviations:
+    MyST: Markedly Structured Text
+  # It is possible to explicitly ignore the `doi-exists` check for certain citation keys
+  error_rules:
+    - rule: doi-exists
+      severity: ignore
+      keys:
+        - Atr03
```
Comment (anwarkhan-ai) [inline papers/rylie_weaver/myst.yml:47]: I believe, error_rules keys are came from the template, and not present in you bib

File: papers/rylie_weaver/mybib.bib [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+	language     = {en},
+	keywords     = {Genome informatics, Machine learning},
+	file         = {Full Text PDF:/Users/r9w/Zotero/storage/EBCU4ED3/Avsec et al. - 2026 - Advancing regulatory variant effect prediction with AlphaGenome.pdf:application/pdf}
```
Comment (anwarkhan-ai) [inline papers/rylie_weaver/mybib.bib]: I suggest avoid your local file path

**Review by anwarkhan-ai — COMMENTED** [review]
@RylieWeaver  I have reviewed, overall its a good package for the SciPy. Like the test suite.

As a reviewer, Is hard to say how model trained, comparing it with the size and time complexity of the published model. 
 A good implementation section is missing that I believe you already have most of it in README file.

Thanks ! I will quickly reply if you need any inputs from me on my review comments.

**Comment by RylieWeaver** [comment]
Hello @anwarkhan-ai,

Thank you for your comments! Since I haven't been able to follow-up / commit the revisions within 24 hours I just wanted to confirm that I've read your comments, cannot immediately see any points of disagreement, and am working on those points. Thank you again for taking your time to review! Will update tomorrow.

File: papers/rylie_weaver/main.md (hunk 5 above)
Comment (RylieWeaver) [reply papers/rylie_weaver/main.md]: Yes I currently host them on my own premium HuggingFace account.

File: papers/rylie_weaver/main.md (hunk 3 above)
Comment (RylieWeaver) [reply papers/rylie_weaver/main.md:64]: @anwarkhan-ai 

Working on this currently. I plan to:
- Show parameter count for (a) the full published architecture from DeepMind and (b) from-scratch models with no heads, one organism, ~1MB max sequence length, and over increasing `num_channels`. The idea is that (a) covers the naive inference/finetuning cases and (b) covers common from-scratch cases.
- Show throughput/memory for the loaded checkpoint architecture over various sequence lengths. We could also do over various `num_channels` sizes, but it might not be worth it the cost of going from a 1D-->2D validation grid.

I'm outlining here first just so you can let me know if I should expand/adjust these methods. Also feel free to wait until they're finished to comment.

**Comment by KandiSanjana** [comment]
Hi! Looks good overall, leaving a few comments 

1. Reference: [main.md, lines 51–67](https://github.com/scipy-conference/scipy_proceedings/blob/c263fe4c0f9043af939bbceab61fe2245dcd835f/papers/rylie_weaver/main.md#L51)
Could you mention the exact alphagenome-pt version or commit used for these examples? The current PyPI release is 0.3.0, while the development branch appears to contain API changes that would make some of these snippets work differently. Pinning the version would help readers reproduce the examples later. If possible, it would also be useful to run the paper’s code snippets in CI against that version.

2. Reference: [main.md, line 35](https://github.com/scipy-conference/scipy_proceedings/blob/c263fe4c0f9043af939bbceab61fe2245dcd835f/papers/rylie_weaver/main.md#L35)
I think “maintaining single base-pair resolution in its predictions” may be slightly too broad. As described later in the paper, some outputs are produced at 1-bp resolution, while others use 128-bp or pairwise resolutions. Would it be more accurate to say that the architecture preserves single-base representations and supports selected predictions at 1-bp resolution?

3. Pairwise representation dimensions
Reference: [main.md, lines 42–44](https://github.com/scipy-conference/scipy_proceedings/blob/c263fe4c0f9043af939bbceab61fe2245dcd835f/papers/rylie_weaver/main.md#L42)
Could you clarify the dimensions of the pairwise representation here? My understanding is that, for a sequence of length S, the spatial shape is approximately [S/2048, S/2048], with each axis representing 2048-bp bins.

4. Reference: [main.md, lines 200–228](https://github.com/scipy-conference/scipy_proceedings/blob/c263fe4c0f9043af939bbceab61fe2245dcd835f/papers/rylie_weaver/main.md#L200)
Could you include the exact Hugging Face repository and checkpoint revision used by this example? Since the checkpoint files are hosted outside the package repository, specifying the repository ID and a revision or release tag would make this workflow more reproducible.

**Comment by RylieWeaver** [comment]
@anwarkhan-ai @KandiSanjana 

We have been working on improving the validation and recently completed full-model JAX–PyTorch equivalence testing (up to numerical thresholds) across multiple precision policies for version 0.4.0. Would it be possible to include these results here and update the proceedings accordingly?

You can see the equivalence tests in the [new documentation](https://alphagenome-pytorch-alphagenome-pt.readthedocs.io/en/latest/deepmind-equivalence.html).

**Comment by RylieWeaver** [comment]
> Hi! Looks good overall, leaving a few comments
> 
> 1. Reference: [main.md, lines 51–67](https://github.com/scipy-conference/scipy_proceedings/blob/c263fe4c0f9043af939bbceab61fe2245dcd835f/papers/rylie_weaver/main.md#L51)
>    Could you mention the exact alphagenome-pt version or commit used for these examples? The current PyPI release is 0.3.0, while the development branch appears to contain API changes that would make some of these snippets work differently. Pinning the version would help readers reproduce the examples later. If possible, it would also be useful to run the paper’s code snippets in CI against that version.
> 2. Reference: [main.md, line 35](https://github.com/scipy-conference/scipy_proceedings/blob/c263fe4c0f9043af939bbceab61fe2245dcd835f/papers/rylie_weaver/main.md#L35)
>    I think “maintaining single base-pair resolution in its predictions” may be slightly too broad. As described later in the paper, some outputs are produced at 1-bp resolution, while others use 128-bp or pairwise resolutions. Would it be more accurate to say that the architecture preserves single-base representations and supports selected predictions at 1-bp resolution?
> 3. Pairwise representation dimensions
>    Reference: [main.md, lines 42–44](https://github.com/scipy-conference/scipy_proceedings/blob/c263fe4c0f9043af939bbceab61fe2245dcd835f/papers/rylie_weaver/main.md#L42)
>    Could you clarify the dimensions of the pairwise representation here? My understanding is that, for a sequence of length S, the spatial shape is approximately [S/2048, S/2048], with each axis representing 2048-bp bins.
> 4. Reference: [main.md, lines 200–228](https://github.com/scipy-conference/scipy_proceedings/blob/c263fe4c0f9043af939bbceab61fe2245dcd835f/papers/rylie_weaver/main.md#L200)
>    Could you include the exact Hugging Face repository and checkpoint revision used by this example? Since the checkpoint files are hosted outside the package repository, specifying the repository ID and a revision or release tag would make this workflow more reproducible.

Firstly, thank you very much for taking the time to review!

On the points:
(1) Would it be possible to update the API calls to v0.4.0? This recent update did sadly change some API calls but it provides a lot better ease-of-use and better JAX-PyTorch equivalence validation (see other comment).
(3) Your understanding is exactly right. After batch and channels dimensions, this becomes [B, S/2048, S/2048, C]
(2/4) Sounds good I will revise for those!

**Comment by RylieWeaver** [comment]
Hello @anwarkhan-ai and @KandiSanjana,

I have updated the proceedings with training curves, numerical-equivalence validation, parameter counts, GPU latency and memory benchmarks, and several wording changes. I also updated material related to features added or revised in v0.4.0, particularly the model calls. I understand that at this stage, revisions should be limited to addressing reviewer comments, so I apologize for the breadth of the update. However, the numerical-equivalence work was substantially expanded for v0.4.0 and directly addresses several comments. Given that addition, it seemed most consistent to update other relevant material to reflect that package version.

I intentionally kept the update in a single commit so that it can be reverted cleanly if you feel it is too broad for this stage. I think the additions, particularly the numerical-equivalence results, substantially strengthen the paper, but I am happy to follow your guidance. Please let me know what you think.

Thank you both again!

**Comment by RylieWeaver** [comment]
@anwarkhan-ai @KandiSanjana 

Just checking in for any final comments before the August 30 author-revision deadline. I’m happy to make any requested revisions or roll back parts of the latest update until then.

Thank you!

**Comment by RylieWeaver** [comment]
Hello @anwarkhan-ai and @KandiSanjana,

I believe the reviewer deadline to make a decision was a couple days ago on September 6th. I didn't receive any notification as of yet; is anything needed from me? Thanks!

**Comment by RylieWeaver** [comment]
Hello @cdlindsey,

Should authors have received a formal decision notification by now, given the September 6 reviewer deadline? If that's coming later, that’s completely fine — I just want to make sure I haven’t missed anything.

I’m tagging you because you're the proceedings chair assigned to my PR, “alphagenome-pt: Flexible Training of AlphaGenome Models in PyTorch.” However, if someone else is better suited to answer, please feel free to tag them.

Thanks!

All the best,
Rylie Weaver

**Comment by anwarkhan-ai** [comment]
Approving. Nice work on the package and on the test suite.

CC @KandiSanjana @cdlindsey

**Review by anwarkhan-ai — APPROVED** [review]
Approved, LGTM.

I have gone through the revised changes, all the review comments are take care.

Good Job!

Thanks !

**Comment by RylieWeaver** [comment]
> Approved, LGTM.
> 
> I have gone through the revised changes, all the review comments are take care.
> 
> Good Job!
> 
> Thanks !

Great, and thank you for your helpful feedback!

All the best,
Rylie

**Comment by KandiSanjana** [comment]
Approved from my end as well! Good work, all the best!

**Comment by RylieWeaver** [comment]
> Approved from my end as well! Good work, all the best!

Great, thank you for your feedback!

Rylie
